components/filter/filter.js

```diff
@@ -0,0 +1,270 @@
+				@d2l-hierarchical-view-show-start="${this._handleDimensionShowStart}"
+				id="${dimension.key}">
+
+				${dimensionHTML}
+			</d2l-hierarchical-view>
+		`;
+	}
+
+	_buildDimensions() {
+		return this._dimensions.map((dimension) => {
+			const builtDimension = this._buildDimension(dimension);
+			return html`<d2l-menu-item text="${dimension.text}">
+				${builtDimension}
+			</d2l-menu-item>`;
+		});
+	}
+
+	_buildHeader() {
+		if (!this._activeDimensionKey) return null;
+
+		const dimensionText = this._dimensions.find(dimension => dimension.key === this._activeDimensionKey).text;
+		return html`
+			<div slot="header">
+				<div class="d2l-filter-dimension-header">
+					<d2l-button-icon
+						@click="${this._handleDimensionHide}"
+						icon="tier1:chevron-left"
+						text="${this.localize('components.menu-item-return.returnCurrentlyShowing', 'menuTitle', dimensionText)}">
+					</d2l-button-icon>
+					<div class="d2l-filter-dimension-header-text d2l-body-standard">${dimensionText}</div>
+				</div>
+			</div>
+		`;
+	}
+
+	_createSetDimension(dimension) {
+		return html`
+			<d2l-list
+				@d2l-list-selection-change="${this._handleChangeSetDimension}"
+				extend-separators>
+
+				${dimension.values.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-set-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_getSlottedNodes(slot) {
+		const nodes = slot.assignedNodes();
+		return nodes.filter((node) => node.nodeType === Node.ELEMENT_NODE && node.tagName.toLowerCase() !== 'template');
+	}
+
+	_handleChangeSetDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-set-value to keep them in sync
+		const dimension = this.querySelector(`d2l-filter-dimension-set[key="${dimensionKey}"`);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = this._getSlottedNodes(slot);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
+
+		this._dispatchChangeEvent({ dimension: dimensionKey, value: { key: valueKey, selected: selected } });
+	}
+
+	_handleDimensionDataChange(e) {
+		const dimension = this._dimensions.find(dimension => dimension.key === e.detail.dimensionKey);
+		const value = e.detail.valueKey && dimension.values.find(value => value.key === e.detail.valueKey);
+		const toUpdate = value ? value : dimension;
+		const changes = e.detail.changes;
+
+		changes.forEach((newValue, prop) => {
+			if (toUpdate[prop] !== newValue) {
+				toUpdate[prop] = newValue;
+				this.requestUpdate();
```

Comment:
May want to only call `requestUpdate` once and not inside a loop. I'm sure it's smart enough to dedupe/wait, but just in case.

Reply:
Yeah I figured it was smart enough, but it was an assumption - I can loop through and then `requestUpdate` at the end if there are any updates

Reply:
I think it should be ok either way - it'll do a single update after the current task.

Reply:
I updated this anyways to make the intent clearer, even if Lit would handle it as expected